src/ffi/mod.rs

```diff
@@ -85,12 +119,604 @@ lazy_static::lazy_static! {
+                        } else if let Some(msg) = info.payload().downcast_ref::<&str>() {
+                            msg.to_string()
+                        } else {
+                            DEFAULT_PANIC_MSG.to_string()
+                        }
+                    };
+                    anyhow::Error::from(panic_handling::Panic { message })
+                };
+
+                // Updating LAST_ERROR.
+                // NOTE: this "could" duplicate updating error, if the error happens on ffi call stack as well ...
+                error_handling::update_last_error(err);
+
+                // Send this "cry for help" to whoever is on the upper side
+                let e = Box::new(
+                    Event::new::<Error>()
+                        .set(ErrorCode::Unknown)
+                        .set(ErrorLevel::Critical)
+                        .set(format!("{}", info)),
+                );
+
+                telio_log_debug!("call_once: {:?}", e);
+                events(e);
+            }));
+        });
+
+        // TODO: Update windows ffi to take in void*, for protect
+        #[cfg(not(target_os = "android"))]
+        let protect = None;
+        #[cfg(target_os = "android")]
+        let protect: Option<Protect> = match protect_cb {
+            Some(protect) => {
+                let protect = protect;
+                Some(Arc::new(move |fd| {
+                    protect.protect(fd);
+                }))
+            }
+            None => None,
+        };
+
+        catch_ffi_panic(|| {
+            let features = features.clone();
+            let event_dispatcher = event_dispatcher.clone();
+            let protect = protect.clone();
+            let device = Device::new(features, event_dispatcher, protect)?;
+            Ok(Self {
+                inner: Mutex::new(device),
+                id: rand::thread_rng().gen::<usize>(),
+            })
+        })
+    }
+
+    fn log_entry(features: Features, serialized_event_fn: String, ret: &Result<telio, TelioError>) {
+        telio_log_info!(
+            "Telio::new entry. features: {:?}. Event_ptr: {serialized_event_fn}. Return value: {:?}",
+            features,
+            ret.as_ref().map(|_| ())
+        );
+    }
+
+    /// Completely stop and uninit telio lib.
+    pub fn destroy(&self) {
+        let mut dev = match self.inner.lock() {
+            Ok(dev) => dev,
+            Err(poisoned) => {
+                telio_log_debug!("main telio lock has been poisoned");
+                poisoned.into_inner()
+            }
+        };
+        dev.stop();
+        dev.shutdown_art();
+    }
+
+    /// Explicitly deallocate telio object and shutdown async rt.
+    pub fn destroy_hard(&self) -> FFIResult<()> {
+        let mut dev = self
+            .inner
+            .lock()
+            .unwrap_or_else(|poisoned| poisoned.into_inner());
+        let res = catch_ffi_panic(|| {
+            dev.try_shutdown(Duration::from_millis(1000))
```

Comment:
Previously `try_shutdown` was destroying `dev` here (and thus deallocating whole object tree starting in `dev`). Why is this no longer done?

Reply:
Because UniFFI doesn't allow you to take ownership of self. It doesn't even allow you to take a mutable reference to self

Reply:
I understand, but why are we no longer destroying that objects/memory - is that still correct behaviour? Maybe `telio` should store represent `inner` as `Mutex<Option<Device>>` and set it to `None` as part of the `try_shutdown`?

Reply:
That sounds like a valid option. I will make the change